### PR TITLE
Translator: Add romanizations setting in Translation settings

### PR DESCRIPTION
**Pull request by eric-p-hutchins**
This adds a new setting in "Translation settings" called "Show romanizations" which tells the translation popup to include in the query the parameter `dt=rm` and then extracts romanizations from the results to display.
translator-romanizations-1
translator-romanizations-2

### CODE REVIEW

File: frontend/ui/translator.lua [hunk 1]
```diff
@@ -277,6 +277,15 @@ This is useful:
                     G_reader_settings:flipNilOrTrue("translator_from_auto_detect")
                 end,
             },
+            {
+                text = _("Show romanizations"),
```
Comment (pazos) [inline frontend/ui/translator.lua:281]: Could you please add also a `help_text` to explain what the feature does in plain english?  :)

I understand it could be useful for all non latin script users¿?

Comment (eric-p-hutchins) [reply frontend/ui/translator.lua:281]: I added it below this. Please review.

**Comment by hius07** [comment]
Can it be just a user patch of 2 lines?

**Comment by eric-p-hutchins** [comment]
> Can it be just a user patch of 2 lines?

I'm sorry I don't understand what you mean by this. 

Are you suggesting I make some change?

**Comment by hius07** [comment]
A niche setting.
This can be solved by a short userpatch
```Lua
local Translator = require("ui/translator")
table.insert(Translator.trans_params.dt, "rm")
```
Userpatches are: https://github.com/koreader/koreader/wiki/User-patches.

**Comment by eric-p-hutchins** [comment]
@hius07 So, I have a few points to respond to that with:

1. It is actually not possible to make that work with user patches. The patch you suggested will add `dt=rm` to the Google Translate API query but it the `Translator:_showTranslation` function still won't know how to extract the romanization since they come in from a different part of the response. It is necessary for the `Translator` module to know about romanizations for this to work, and there's nothing for the user to modify to hack in support for it.
2. The point is also to be able to conveniently enable/disable it whenever the user wants. It may be more needed for some content than for others. I don't see examples in userpatches of adding to the settings, which is an important part of the change.
3. It's not clear to me what the reasoning is for whether something should be merged vs. a userpatch. If transliteration is niche, then why is translation itself not niche as well? They show up on the Google Translate UI for a reason, after all. It is not obvious that userpatches even exist, let alone how to install them.
4. Basic support for the Google Translate API has been in place for years and it is even documented in the comments that this query parameter exists to get transliterations, but it has not been used. I think this is a natural extension of the `Translator` interface, providing an easily added feature that makes a big difference but is, in fact, not trivial to add without modifying the internals of the module.
5. The "Translation settings" menu only has 4 items so far. Some settings have pages of items to scroll through. Going from 4 items to 5 seems like not a big deal imo.
6. If the reasoning for not merging is maintenance burden then I don't see how this code causes any issues with existing functionality. The code already has to be defensive about extracting the real translations since the response arrays may or may not have real results in all the slots, so it checks for string results. Romanizations show up in a different part of the response, so it will never conflict with where it checks for regular translations and vice versa.

**Comment by hius07** [comment]
> 1. It is necessary for the `Translator` module to know about romanizations for this to work

Good point, agree.

File: frontend/ui/translator.lua [hunk 2]
```diff
@@ -377,6 +387,20 @@ function Translator:loadPage(text, target_lang, source_lang)
...
+    if G_reader_settings:isTrue("translator_from_romanizations") then
+       table.insert(self.trans_params.dt, "rm")
+    else
+       local rmIndex = nil
+       for i, v in pairs(self.trans_params.dt) do
+          if v == "rm" then
+             rmIndex = i
+             break
+          end
+       end
```
Comment (hius07) [inline frontend/ui/translator.lua:399]: We have
https://github.com/koreader/koreader/blob/8297d17e54374fcb568e14b03a3a9ca94aac577e/frontend/util.lua#L323

**Comment by hius07** [comment]
While we are in Translator.
https://github.com/koreader/koreader/blob/8297d17e54374fcb568e14b03a3a9ca94aac577e/frontend/ui/translator.lua#L296
@Frenzie, isn't it more correct `Translate into:`?

**Comment by Frenzie** [comment]
I think it's a from/to, but into is also fine.

File: frontend/ui/translator.lua [hunk 3]
```diff
@@ -277,6 +277,16 @@ This is useful:
...
+                checked_func = function()
+                    return G_reader_settings:isTrue("translator_from_romanizations")
+                end,
+                callback = function()
+                    G_reader_settings:flipTrue("translator_from_romanizations")
+                end,
+            },
             {
                 text_func = function()
                     local lang = G_reader_settings:readSetting("translator_from_language")
```
Comment (poire-z) [inline frontend/ui/translator.lua:292]: (May be you could name the setting `translator_with_romanizations` or `translator_show_romanizations` to avoivd "same prefix but different purpose" with `translator_from_language` with using `from`?)

File: frontend/ui/translator.lua [hunk 4]
```diff
@@ -377,6 +387,20 @@ function Translator:loadPage(text, target_lang, source_lang)
     local query = ""
     self.trans_params.tl = target_lang
     self.trans_params.sl = source_lang
+    if G_reader_settings:isTrue("translator_from_romanizations") then
+       table.insert(self.trans_params.dt, "rm")
+    else
+       local rmIndex = nil
+       for i, v in pairs(self.trans_params.dt) do
+          if v == "rm" then
+             rmIndex = i
```
Comment (poire-z) [inline frontend/ui/translator.lua:396]: No other way avoid this and the ugly removal? Dunno (no time to check) how this appears in the query when dt is an array.

Comment (eric-p-hutchins) [reply frontend/ui/translator.lua]: Hmm, you're right. 

Actually, I only had to do this to make the tests that I added pass since it reuses the `Translator` between tests. Actually, I'm pretty sure from testing it for real that this doesn't need to be done otherwise. 

Do you have a suggestion how to better handle using this in tests?

File: frontend/ui/translator.lua (hunk 3 above)
Comment (eric-p-hutchins) [reply frontend/ui/translator.lua:292]: Yeah, I re-used the `_from_` segment to mean that the romanizations come from the "from" language (rather than the "to" language).

But maybe that's a little confusing.

File: frontend/ui/translator.lua (hunk 4 above)
Comment (eric-p-hutchins) [reply frontend/ui/translator.lua]: I moved the initialization of `dt` to this function so that it doesn't affect the tests. Now the tests still pass and it's not as ugly. Thanks for asking for me to look at this again.

File: frontend/ui/translator.lua (hunk 3 above)
Comment (eric-p-hutchins) [reply frontend/ui/translator.lua]: Renamed to `translator_with_romanizations`

File: frontend/ui/translator.lua [hunk 5]
```diff
@@ -188,19 +188,6 @@ local Translator = {
         ssel = 0,  -- ?
         tsel = 0,  -- ?
         -- tk = "" -- auth token
-        dt = { -- what we want in result
-            "t",   -- translation of source text
-            "at",  -- alternate translations
-            -- Next options only give additional results when text is a single word
```
[the remark this answers is not in the document]
Comment (eric-p-hutchins) [reply frontend/ui/translator.lua:194]: That's probably why I ended up with the approach I had at first :sweat_smile: 

I can put the comments back here but I still think that `dt` should be added later on rather than here, so that it is clearer what's happening later on.

[the remark this answers is not in the document]
Comment (poire-z) [reply frontend/ui/translator.lua]: We could put them all (with their documentation) where there are used, but that's a bit more than what this PR is about :)

Why there are here may not have been really thought (I may have just started translator.lua by copying wikipedia.lua), but the side effect of having it there is that people can user-patch it (like we suggested you to do :)) - so by hardcoded the replacement in the body of a function, the user patched stuff will be overridden.

File: frontend/ui/translator.lua [hunk 6]
```diff
@@ -377,6 +374,23 @@ function Translator:loadPage(text, target_lang, source_lang)
...
+       -- "rm",  -- transcription / transliteration of source and translated texts
+       -- "ss",  -- synonyms of source text, if it's one word
+    }
+
+    if G_reader_settings:isTrue("translator_with_romanizations") then
+       table.insert(self.trans_params.dt, "rm")
+    end
     for k,v in pairs(self.trans_params) do
         if type(v) == "table" then
             for _, v2 in ipairs(v) do
```
[the remark this answers is not in the document]
Comment (poire-z) [reply frontend/ui/translator.lua:397]: I don't mind the special case :)
~~But if you really don't want that, you could:
```lua
    local dt = self.trans_params.dt
    if G_reader_settings:isTrue("translator_with_romanizations") then
       dt = util.tableDeepCopy(dt)
       table.insert(dt, "rm"}
    end
```
Won't work as we just process the full self.trans_params...
I don't see another clean way.

File: frontend/ui/translator.lua [hunk 7]
```diff
@@ -188,19 +188,19 @@ local Translator = {
...
+        -- dt = { -- what we want in result
+            -- "t",   -- translation of source text
+            -- "at",  -- alternate translations
             -- Next options only give additional results when text is a single word
             -- "bd",  -- dictionary (articles, reverse translations, etc)
             -- "ex",  -- examples
             -- "ld",  -- ?
-            "md",  -- definitions of source text
+            -- "md",  -- definitions of source text
             -- "qca", -- ?
```
[the remark this answers is not in the document]
Comment (eric-p-hutchins) [reply frontend/ui/translator.lua]: Ok, I put that back.

File: frontend/ui/translator.lua (hunk 6 above)
[the remark this answers is not in the document]
Comment (eric-p-hutchins) [reply frontend/ui/translator.lua]: Alright, I went with appending to the query directly like you suggested.

**Review by poire-z — APPROVED** [review]

**Review by Frenzie — APPROVED** [review]

**Review by pazos — APPROVED** [review]
